File: packages/health_connector_hk_ios/test/unit_tests/src/mappers/health_record_mappers/vo2_max_record_mapper_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hk_ios/src/mappers/health_record_mappers/vo2_max_record_mapper.dart';
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart';

import '../../../utils/fake_data.dart';

void main() {
  group(
    'Vo2MaxRecordMapper',
    () {
      group(
        'Vo2MaxRecordToDto',
        () {
          test(
            'converts Vo2MaxRecord to Vo2MaxRecordDto',
            () {
              final time = FakeData.fakeTime;

              final record = Vo2MaxRecord(
                id: HealthRecordId(FakeData.fakeId),
                time: time,
                zoneOffsetSeconds: FakeData.fakeTime.timeZoneOffset.inSeconds,
                metadata: const Metadata(
                  dataOrigin: DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.activelyRecorded,
                  clientRecordVersion: 1,
                  device: Device(type: DeviceType.watch),
                ),
                mLPerKgPerMin: const Number(45.5),
              );

              final dto = record.toDto();

              expect(dto.id, FakeData.fakeId);
              expect(dto.time, time.millisecondsSinceEpoch);
              expect(
                dto.zoneOffsetSeconds,
                FakeData.fakeTime.timeZoneOffset.inSeconds,
              );
              expect(dto.mLPerKgPerMin.value, 45.5);
              expect(dto.metadata.dataOrigin, FakeData.fakeDataOrigin);
            },
          );
        },
      );

      group(
        'Vo2MaxRecordDtoToDomain',
        () {
          test(
            'converts Vo2MaxRecordDto to Vo2MaxRecord',
            () {
              final time = FakeData.fakeTime;

              final dto = Vo2MaxRecordDto(
                id: FakeData.fakeId,
                time: time.millisecondsSinceEpoch,
                zoneOffsetSeconds: FakeData.fakeTime.timeZoneOffset.inSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.activelyRecorded,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.watch,
                ),
                mLPerKgPerMin: NumberDto(value: 42.3),
              );

              final record = dto.toDomain();

              expect(record.id.value, FakeData.fakeId);
              expect(record.time, time);
              expect(record.mLPerKgPerMin.value, 42.3);
              expect(
                record.metadata.dataOrigin.packageName,
                FakeData.fakeDataOrigin,
              );
            },
          );

          test(
            'converts Vo2MaxRecordDto with null id to domain with none id',
            () {
              final time = FakeData.fakeTime;

              final dto = Vo2MaxRecordDto(
                time: time.millisecondsSinceEpoch,
                zoneOffsetSeconds: FakeData.fakeTime.timeZoneOffset.inSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                mLPerKgPerMin: NumberDto(value: 38.0),
              );

              final record = dto.toDomain();

              expect(record.id, HealthRecordId.none);
            },
          );
        },
      );
    },
  );
}

File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/Vo2MaxRecordMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.records.Vo2MaxRecord
import androidx.health.connect.client.records.metadata.Metadata
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import com.phamtunglam.health_connector_hc_android.pigeon.Vo2MaxMeasurementMethodDto
import com.phamtunglam.health_connector_hc_android.pigeon.Vo2MaxRecordDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import java.time.ZoneOffset
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("Vo2MaxRecordMapper")
class Vo2MaxRecordMapperTest {

    private companion object {
        const val TEST_ID = "vo2max-id"
        const val TEST_TIME = 1609459200000L
        const val TEST_VO2 = 45.0
        val TEST_ZONE_OFFSET: ZoneOffset = ZoneOffset.ofHours(1)
    }

    @Test
    @DisplayName("GIVEN Vo2MaxRecord → WHEN toDto → THEN converts correctly")
    fun whenRecordToDto_thenConvertsCorrectly() {
        // Given
        val record = Vo2MaxRecord(
            time = Instant.ofEpochMilli(TEST_TIME),
            zoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
            vo2MillilitersPerMinuteKilogram = TEST_VO2,
            measurementMethod = Vo2MaxRecord.MEASUREMENT_METHOD_COOPER_TEST,
        )

        // When
        val result = record.toDto()

        // Then
        result.millilitersPerKilogramPerMinute shouldBe TEST_VO2
        result.time shouldBe TEST_TIME
        result.zoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.measurementMethod shouldBe Vo2MaxMeasurementMethodDto.COOPER_TEST
    }

    @Test
    @DisplayName("GIVEN Vo2MaxRecordDto → WHEN toHealthConnect → THEN creates correctly")
    fun whenDtoToHealthConnect_thenCreatesCorrectly() {
        // Given
        val dto = Vo2MaxRecordDto(
            id = TEST_ID,
            time = TEST_TIME,
            zoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            millilitersPerKilogramPerMinute = TEST_VO2,
            measurementMethod = Vo2MaxMeasurementMethodDto.METABOLIC_CART,
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.vo2MillilitersPerMinuteKilogram shouldBe TEST_VO2
        result.time shouldBe Instant.ofEpochMilli(TEST_TIME)
        result.zoneOffset shouldBe TEST_ZONE_OFFSET
        result.measurementMethod shouldBe Vo2MaxRecord.MEASUREMENT_METHOD_METABOLIC_CART
        result.metadata.id shouldBe TEST_ID
    }
}
